spatialdata/_core/models.py

```diff
@@ -0,0 +1,421 @@
+    name = list(name)[0]
+    for k in data.keys():
+        schema.validate(data[k][name])
+    return data
+
+
+# TODO: should check for column be strict?
+# TODO: validate attrs for transform.
+Polygons_s = DataFrameSchema(
+    {
+        "geometry": Column(GeometryDtype),
+    }
+)
+
+
+@singledispatch
+def validate_polygons(data: Any, *args: Any, **kwargs: Any) -> GeoDataFrame:
+    """
+    Validate (or parse) polygons data.
+
+    Parameters
+    ----------
+    data
+        Data to validate.
+    offsets
+        TODO.
+    geometry
+        If 3 is :class:`shapely.geometry.Polygon`, if 6 is :class:`shapely.geometry.MultiPolygon`.
+    transform
+        Transformation to apply to the data.
+
+    Returns
+    -------
+    :class:`geopandas.geodataframe.GeoDataFrame`.
+    """
+    raise ValueError(f"Unsupported type: {type(data)}")
+
+
+@validate_polygons.register
+def _(
+    data: np.ndarray,  # type: ignore[type-arg]
+    offsets: Tuple[np.ndarray, ...],  # type: ignore[type-arg]
+    geometry: Literal[3, 6],
+    transform: Optional[Any] = None,
+    **kwargs: Any,
+) -> GeoDataFrame:
+
+    geometry = GeometryType(geometry)
+
+    data = from_ragged_array(geometry, data, offsets)
+    geo_df = GeoDataFrame({"geometry": data})
+    if transform is None:
+        transform = Identity()
+    geo_df.attrs = {"transform": transform}
+    Polygons_s.validate(geo_df)
+    return geo_df
+
+
+@validate_polygons.register
+def _(
+    data: str,
+    transform: Optional[Any] = None,
+    **kwargs: Any,
+) -> GeoDataFrame:
+
+    data = from_geojson(data)
+    geo_df = GeoDataFrame({"geometry", data})
+    if transform is None:
+        transform = Identity()
+    geo_df.attrs = {"transform": transform}
+    Polygons_s.validate(geo_df)
+    return geo_df
+
+
+@validate_polygons.register
+def _(
+    data: GeoDataFrame,
+    transform: Optional[Any] = None,
+    *args: Any,
+    **kwargs: Any,
+) -> GeoDataFrame:
+
+    Polygons_s.validate(data)
+    if transform is None:
+        transform = Identity()
+    data.attrs = {"transform": transform}
+    return data
+
+
+# TODO: add schema for validation.
+@singledispatch
+def validate_shapes(data: Any, *args: Any, **kwargs: Any) -> AnnData:
+    """
+    Validate (or parse) shapes data.
+
+    Parameters
+    ----------
+    data
+        Data to validate.
+    shape_type
+        Type of shape to validate. Can be "Circle" or "Square".
+    shape_size
+        Size of shape to validate.
+
+    Returns
+    -------
+    :class:`anndata.AnnData`.
+    """
+    raise ValueError(f"Unsupported type: {type(data)}")
+
+
+@validate_shapes.register
+def _(
+    data: np.ndarray,  # type: ignore[type-arg]
+    shape_type: Literal["Circle", "Square"],
+    shape_size: np.float_,
+    transform: Optional[Any] = None,
+    **kwargs: Any,
+) -> AnnData:
+
+    # TODO: AnnData(obsm={"spatial": data}) doesn't work, shall we change?
+    adata = AnnData(np.empty(shape=data.shape), obsm={"spatial": data})
```

Comment:
@ivirshup this `AnnData(obsm={"spatial": data})` doesn't work. Quite important if we want to save shapes/points in anndata with coordinates in obsm (and X potentially `None` ). What should we do?